### PR TITLE
The `--empty` flag

### LINKED ISSUE #4572: Support --empty flag for schema-only dry runs from dbt-core PR #8971
This change includes a variety of changes to improve the functionality and usability of the codebase. The most important changes include adding new test cases for running `dbt run` with and without the `--empty` flag, adding a new optional `limit` parameter to the `create_ephemeral_from_node` method and the `BaseRelation` class, and adding a new command line option `--empty` to limit the input refs and sources to zero rows.

### Solution
* `BaseRelation` changes
  * Adds an optional `limit` to `BaseRelation`
  * Implements `BaseRelation.render_limited`: depending on the value of `self.limit`, this method templates sql to wrap a rendered relation in an 'empty' select statement. Further reasoning in the[ spike report](https://github.com/dbt-labs/dbt-core/pull/8971#issuecomment-1789596936)
  * Updates `BaseRelation.__str__` to use `self.render_limited` if `self.limit` is set
* Adds `--empty` flag to `run` and `build` commands.
   * This flag configuration is used to determine the value for `resolve_limit` on the `BaseResolver`
   * `BaseResolver` plumbs its `resolve_limit` to the `Relation` instance it is creating, which is responsible for rendering itself given an optional `limit` parameter.

### PR DESCRIPTION
**Pull request by matthewshaver**
## What are you changing in this pull request and why?

Updates the `run` and `build` command references to include the `--empty` flag.

## Checklist

Uncomment when publishing docs for a prerelease version of dbt:
- [x] Add versioning components, as described in [Versioning Docs](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-entire-pages)
- [x] Add a note to the prerelease version [Migration Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)

- [x] Review the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) so my content adheres to these guidelines.
- [x] For [docs versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#about-versioning), review how to [version a whole page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) and [version a block of content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content).

### CODE REVIEW

File: website/docs/docs/dbt-versions/core-upgrade/01-upgrading-to-v1.8.md [hunk 1]
```diff
@@ -28,7 +28,12 @@ dbt Labs is committed to providing backward compatibility for all versions 1.x,
 
 ## New and changed features and functionality
 
-Coming soon!
+Features and functionality new in dbt v1.8
+
+### The `--empty` flag
+
+The [`run`](/reference/commands/run#the-`--empty`-flag) and [`build`](/reference/commands/run#the-`--empty`-flag) commands now support the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will build the project and models, parse the files to identify changes, but won't actually execute the SQL against the target data warehouse. This ensures your models will build properly and validates dependencies.
```
Comment (MichelleArk) [inline website/docs/docs/dbt-versions/core-upgrade/01-upgrading-to-v1.8.md]: > dbt will build the project and models, parse the files to identify changes, but won't actually execute the SQL against the target data warehouse.

I don't think this is technically correct. Perhaps: 
"dbt will still execute the model SQL against the target data warehouse, but will avoid expensive reads of input data. This ensures your models will build properly and validates dependencies."

Comment (matthewshaver) [inline website/docs/docs/dbt-versions/core-upgrade/01-upgrading-to-v1.8.md]:
```suggestion
The [`run`](/reference/commands/run#the-`--empty`-flag) and [`build`](/reference/commands/run#the-`--empty`-flag) commands now support the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will still execute the model SQL against the target data warehouse but will avoid expensive reads of input data. This validates dependencies and ensures your models will build properly.
```

File: website/docs/reference/commands/build.md [hunk 2]
```diff
@@ -25,6 +25,14 @@ In DAG order, for selected resources or an entire project.
 
 **Flags:** The `build` task supports all the same flags as `run`, `test`, `snapshot`, and `seed`. For flags that are shared between multiple tasks (e.g. `--full-refresh`), `build` will use the same value for all selected resource types (e.g. both models and seeds will be full refreshed).
 
+<VersionBlock firstVersion="1.8">
+
+### The `--empty` flag
+
+The `build` command supports the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will build the project and models, parse the files to identify changes, but won't actually execute the SQL against the target data warehouse. This ensures your models will build properly and validates dependencies.
```
Comment (matthewshaver) [inline website/docs/reference/commands/build.md]:
```suggestion
The `build` command supports the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will still execute the model SQL against the target data warehouse but will avoid expensive reads of input data. This validates dependencies and ensures your models will build properly.
```

File: website/docs/reference/commands/run.md [hunk 3]
```diff
@@ -80,3 +80,11 @@ See [global configs](/reference/global-configs/failing-fast)
 ## Enable or Disable Colorized Logs
 
 See [global configs](/reference/global-configs/print-output#print-color)
+
+<VersionBlock firstVersion="1.8">
+
+## The `--empty` flag
+
+The `run` command supports the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will build the project and models, parse the files to identify changes, but won't actually execute the SQL against the target data warehouse. This ensures your models will build properly and validates dependencies.
```
Comment (matthewshaver) [inline website/docs/reference/commands/run.md]:
```suggestion
The `run` command supports the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will still execute the model SQL against the target data warehouse but will avoid expensive reads of input data. This validates dependencies and ensures your models will build properly.
```

**Review by MichelleArk — APPROVED** [review]

**Review by mirnawong1 — APPROVED** [review]
lgtm and since it's approved already, only non blocker that i think will be helpful for users is adding a 'why is this useful or important in the 'build' and 'run' pages:


> This validates dependencies and ensures your models will build properly. Using the `--empty` flag enhances performance and helps with warehouse costs (not sure if that's right?)

File: website/docs/docs/dbt-versions/core-upgrade/01-upgrading-to-v1.8.md [hunk 4]
```diff
@@ -28,7 +28,12 @@ dbt Labs is committed to providing backward compatibility for all versions 1.x,
 
 ## New and changed features and functionality
 
-Coming soon!
+Features and functionality new in dbt v1.8
+
+### The `--empty` flag
+
+The [`run`](/reference/commands/run#the-`--empty`-flag) and [`build`](/reference/commands/run#the-`--empty`-flag) commands now support the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will still execute the model SQL against the target data warehouse but will avoid expensive reads of input data. This validates dependencies and ensures your models will build properly.
```
Comment (mirnawong1) [inline website/docs/docs/dbt-versions/core-upgrade/01-upgrading-to-v1.8.md]: i think the link is wrong matt 

```suggestion
The [`run`](/reference/commands/run#the-`--empty`-flag) and [`build`](/reference/commands/build#the---empty-flag) commands now support the `--empty` flag for building schema-only dry runs. The `--empty` flag limits the refs and sources to zero rows. dbt will still execute the model SQL against the target data warehouse but will avoid expensive reads of input data. This validates dependencies and ensures your models will build properly.
```

Comment (matthewshaver) [reply website/docs/docs/dbt-versions/core-upgrade/01-upgrading-to-v1.8.md]: Good catch!
